ciao-controller/client.go

```
@@ -114,26 +109,84 @@ func (client *ssntpClient) releaseResources(instanceID string) error {
 	return nil
 }
 
-func (client *ssntpClient) removeInstance(instanceID string) {
-	err := client.releaseResources(instanceID)
+func (client *ssntpClient) removeInstance(i *types.Instance) {
+	err := client.releaseResources(i)
 	if err != nil {
 		glog.Warningf("Error when releasing resources for deleted instance: %v", err)
 	}
-	client.deleteEphemeralStorage(instanceID)
-	err = client.ctl.ds.DeleteInstance(instanceID)
+	client.deleteEphemeralStorage(i.ID)
+	err = client.ctl.ds.DeleteInstance(i.ID)
 	if err != nil {
 		glog.Warningf("Error deleting instance from datastore: %v", err)
 	}
+
+	if i.CNCI {
+		return
+	}
+
+	// check the cnci after 5 minutes has elapsed to see if it's idle.
+	go func() {
+		time.Sleep(5 * time.Minute)
+		err = client.checkCNCI(i.TenantID)
+		if err != nil {
+			glog.Warningf("Error deleting CNCI: %v", err)
+		}
+	}()
+}
+
+func (client *ssntpClient) checkCNCI(tenantID string) error {
+	t, err := client.ctl.ds.GetTenant(tenantID)
+	if err != nil {
+		return errors.Wrapf(err, "error getting tenant from datastore")
+	}
+
+	cncis, err := client.ctl.ds.GetTenantCNCISummary(t.CNCIID)
+	if err != nil {
+		return errors.Wrapf(err, "error getting cnci from datastore")
+	}
+
+	// tenants only have one cnci right now
+	cnci := cncis[0]
+
+	if len(cnci.Subnets) > 0 {
+		return nil
+	}
+
+	// cache this value.
+	cnciID := t.CNCIID
+
+	// clean up tenant info in datastore.
+	err = client.ctl.ds.RemoveTenantCNCI(t.ID)
+	if err != nil {
+		// probably this is ok, it means that something needs
```

Let's have a specific error for that case, just silently ignoring errors has been a source of bugs in the past.